src/Mod/TechDraw/Gui/SingleInsertTool.cpp

```diff
@@ -0,0 +1,215 @@
+        bool is_linked = false;
+        auto obj = sel.getObject();
+        if (obj->isDerivedFrom<TechDraw::DrawPage>() || obj->isDerivedFrom<TechDraw::DrawView>()) {
+            continue;
+        }
+
+        if (obj->isDerivedFrom<Spreadsheet::Sheet>()) {
+            Gui::Command::openCommand(QT_TRANSLATE_NOOP("Command", "Create BIM View"));
+            ViewMakers::makeSpreadsheetView(page);
+            viewCreated = true;
+            Gui::Command::commitCommand();
+            continue;
+        }
+
+        if (DrawGuiUtil::isArchSection(obj)) {
+            Gui::Command::openCommand(QT_TRANSLATE_NOOP("Command", "Create BIM View"));
+            ViewMakers::makeBIMView(page);
+            Gui::Command::commitCommand();
+            viewCreated = true;
+            continue;
+        }
+
+        if (obj->isDerivedFrom<App::LinkElement>()
+            || obj->isDerivedFrom<App::LinkGroup>()
+            || obj->isDerivedFrom<App::Link>()) {
+            is_linked = true;
+        }
+        // If parent of the obj is a link to another document, we possibly need to treat non-link obj as linked, too
+        // 1st, is obj in another document?
+        if (obj->getDocument() != cmd.getDocument()) {
+            std::set<App::DocumentObject*> parents = obj->getInListEx(true);
+            for (auto& parent : parents) {
+                // Only consider parents in the current document, i.e. possible links in this View's document
+                if (parent->getDocument() != cmd.getDocument()) {
+                    continue;
+                }
+                // 2nd, do we really have a link to obj?
+                if (parent->isDerivedFrom<App::LinkElement>()
+                    || parent->isDerivedFrom<App::LinkGroup>()
+                    || parent->isDerivedFrom<App::Link>()) {
+                    // We have a link chain from this document to obj, and obj is in another document -> it is an XLink target
+                    is_linked = true;
+                }
+            }
+        }
+        if (is_linked) {
+            xShapes.push_back(obj);
+            continue;
+        }
+        //not a Link and not null.  assume to be drawable.  Undrawables will be
+        // skipped later.
+        shapes.push_back(obj);
+        if (partObj) {
+            continue;
+        }
+        //don't know if this works for an XLink
+        for (auto& sub : sel.getSubNames()) {
+            if (TechDraw::DrawUtil::getGeomTypeFromName(sub) == "Face") {
+                faceName = sub;
+                //
+                partObj = obj;
+                break;
+            }
+        }
+    }
+
+    if (shapes.empty() &&
+        xShapes.empty() &&
+        !viewCreated) {
+        if (Preferences::ShowInsertFileMessage()) {
+            auto msgText = QObject::tr("If you want to insert a view from existing objects, please select them before invoking this tool. Without a selection, a file browser will open, to insert a SVG or image file.");
```

Comment:
Are this meant to be preprended with `QObject`, wouldn't we loose translation context? I don't know myself